### PR TITLE
0.2 dev

### PR DESCRIPTION
**Pull request by groovili**

### CODE REVIEW

File: DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -28,6 +28,8 @@ public function getConfigTreeBuilder()
                     ->children()
                         ->arrayNode('email')
                             ->children()
+                                ->variableNode('from')
```
Comment (kopaygorodsky) [inline DependencyInjection/Configuration.php:31]: what if I want to send messages from no-reply/info/accounts/support@gmail.com? No reason to keep only one. 
I think user can manage it himself.

File: Service/NotificationManager.php [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      return $notification;
+    }
+
+    /**
+     * @param \Kopay\NotificationBundle\Entity\Notification $notification
+     * @param array $options
+     */
+    public function send(Notification $notification, array $options)
+    {
+      // TODO: Implement creation of job dependent on notification type. In
```
Comment (kopaygorodsky) [inline Service/NotificationManager.php]: $db->dispatch('created', new event($notification) here

File: Service/NotificationManager.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+      if(!in_array($type, self::NOTIFICATION_TYPE)){
+        throw new InvalidArgumentException('Unknown type of notification.');
+      }
+
+      $title = (isset($options['title'])) ? $options['title'] : '';
+      $message = (isset($options['message'])) ? $options['message'] : '';
+
+      switch ($type) {
+        case self::NOTIFICATION_TYPE['Push']:
```
Comment (kopaygorodsky) [inline Service/NotificationManager.php]: Why I need to type 'Push'? 
I suggest 
const NOTIFICATION_PUSH
const NOTIFICATION_EMAIL

**Review by kopaygorodsky — CHANGES_REQUESTED** [review]

File: Entity/NotificationMessageInterface.php [hunk 4]
```diff
@@ -1,11 +1,10 @@
 <?php
 
-namespace Kopay\NotificationBundle\Entity;
+  namespace Kopay\NotificationBundle\Entity;
 
-use Doctrine\Common\Collections\Collection;
+  use Doctrine\Common\Collections\Collection;
```
Comment (akovalyov) [inline Entity/NotificationMessageInterface.php:5]: Please remove. array as a return type is totally sufficient.

File: Entity/NotificationMessageInterface.php [hunk 5]
```diff
@@ -1,11 +1,10 @@
...
 
-interface NotificationMessageInterface
-{
+  interface NotificationMessageInterface {
```
Comment (akovalyov) [inline Entity/NotificationMessageInterface.php:7]: brace to the new line please. Let's follow PSR-2.

File: Entity/NotificationMessageInterface.php [hunk 6]
```diff
@@ -21,9 +20,19 @@ public function getRecipientsItems(): Collection;
     public function getMessage(): ? string;
 
     /**
+     * @param null|string $message
+     */
+    public function setMessage(?string $message): void;
```
Comment (akovalyov) [inline Entity/NotificationMessageInterface.php:25]: can message be nullable? What is the usecase then?

File: Event/NotificationEventInterface.php [hunk 7]
```diff
@@ -6,12 +6,12 @@
...
 interface NotificationEventInterface
 {
-    const NOTIFICATION_CREATED = 'kopaygorodsky.notification.message.created';
-    const NOTIFICATION_POST_PERSIST = 'kopaygorodsky.notification.message.postPersist';
-    const JOB_CREATED = 'kopaygorodsky.notification.job.created';
-    const JOB_PRE_SEND = 'kopaygorodsky.notification.job.preSend';
-    const JOB_POST_SEND = 'kopaygorodsky.notification.job.postSend';
-    const JOB_FAILED = 'kopaygorodsky.notification.job.failed';
+    const NOTIFICATION_CREATED = 'kopay.notify.message.created';
+    const NOTIFICATION_POST_PERSIST = 'kopay.notify.message.postPersist';
```
Comment (akovalyov) [inline Event/NotificationEventInterface.php:10]: symfony convention is snake case for event names, example - https://github.com/symfony/form/blob/master/FormEvents.php#L34

File: Service/NotificationManager.php [hunk 8]
```diff
@@ -0,0 +1,170 @@
+<?php
+  /**
```
Comment (akovalyov) [inline Service/NotificationManager.php:2]: it's a place for license if any. Better place to declare your authority is above the class declaration. There is a special [`@author` annotation for that](http://docs.phpdoc.org/references/phpdoc/tags/author.html).

File: Service/NotificationManager.php [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+  namespace Kopay\NotificationBundle\Service;
+
+  use Kopay\NotificationBundle\Entity\NotificationEmail;
+  use Kopay\NotificationBundle\Entity\NotificationMessageInterface;
+  use Kopay\NotificationBundle\Entity\NotificationPush;
+  use Symfony\Component\OptionsResolver\Exception\InvalidArgumentException;
+
+  /**
+   * Class NotificationManager
```
Comment (akovalyov) [inline Service/NotificationManager.php:16]: redundant.

File: Service/NotificationManager.php [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+   *
+   * @package Kopay\NotificationBundle\Service
```
Comment (akovalyov) [inline Service/NotificationManager.php:18]: redundant.

File: Service/NotificationManager.php [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+   */
+  class NotificationManager {
```
Comment (akovalyov) [inline Service/NotificationManager.php:20]: What does this class actually manage? 
This class violates SRP - it behaves like a factory for notifications, it sends something and it provides a fluent interface to create notifications - this should be at least 3 separate classes - NotificationFactory, NotificationBuider and NotificationSender.

File: Service/NotificationManager.php [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * @var array
+     */
+    private $configuration;
+
+    protected const NOTIFICATION_EMAIL = 0;
```
Comment (akovalyov) [inline Service/NotificationManager.php:26]: Why 0/1 as constant values? What if SMS will be next notification, should we add magic 3? Please use meaningful string values - they are more verbose.

File: Service/NotificationManager.php [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+    protected const NOTIFICATION_PUSH  = 1;
+
+    /**
+     * @var NotificationMessageInterface
+     */
+    protected $notification = null;
+
+    /**
+     * NotificationManager constructor.
```
Comment (akovalyov) [inline Service/NotificationManager.php:36]: redundant.

File: Service/NotificationManager.php [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @param array $config
+     */
+    public function __construct(array $config)
+    {
+      $this->configuration = $config;
```
Comment (akovalyov) [inline Service/NotificationManager.php:42]: no validation? So, if I pass empty array here - what will happen? I would suggest to flattern this array and make each required key as an argument here.

File: Service/NotificationManager.php [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        throw new InvalidArgumentException('Unknown type of notification.');
+      }
+
+      switch ($type) {
+        case self::NOTIFICATION_EMAIL:
+          $this->notification = new NotificationEmail();
+          break;
+        case self::NOTIFICATION_PUSH:
+          $this->notification = new NotificationPush();
+          break;
```
Comment (akovalyov) [inline Service/NotificationManager.php:62]:
```php
default:
    throw new \InvalidArgumentException(sprintf('Case `%s` was not implemented`, $type);
```

File: Service/NotificationManager.php [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+      }
+
+      return $this;
```
Comment (akovalyov) [inline Service/NotificationManager.php:65]: no fluent interface please.

File: Service/NotificationManager.php [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    /**
+     * @param string $title
+     *
+     * @return \Kopay\NotificationBundle\Service\NotificationManager|null
+     */
+    public function setTitle(string $title): ?NotificationManager
```
Comment (akovalyov) [inline Service/NotificationManager.php:73]: shouldn't user construct the message himself and then just send it out? Why do we do that for him?

File: Service/NotificationManager.php [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * @param string|null $from
+     *
+     * @return \Kopay\NotificationBundle\Service\NotificationManager|null
+     */
+    public function setFrom(string $from = null): ?NotificationManager
+    {
+      if (!($this->notification instanceof NotificationEmail)) {
+        throw new InvalidArgumentException('Can\'t set from to push notification.');
```
Comment (akovalyov) [inline Service/NotificationManager.php:100]: should be in the entity itself.

File: Service/NotificationManager.php [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+      }
+
+      if (is_null($from)) {
```
Comment (akovalyov) [inline Service/NotificationManager.php:103]: `null === $from`

File: Service/NotificationManager.php [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * @param array $recipients
+     *
+     * @return \Kopay\NotificationBundle\Service\NotificationManager|null
+     */
+    public function setRecipients(array $recipients): ?NotificationManager
+    {
+      if (empty($recipients)) {
+        throw new InvalidArgumentException('At least one recipient required.');
```
Comment (akovalyov) [inline Service/NotificationManager.php:120]: you're not referring to global namespace. Do we have an own `InvalidArgumentException` implementation?

File: Service/NotificationManager.php [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * @param array $value
+     *
+     * @return \Kopay\NotificationBundle\Service\NotificationManager|null
+     */
+    public function setValue(array $value): ?NotificationManager
+    {
+      if (!($this->notification instanceof NotificationPush)) {
+        throw new InvalidArgumentException('Can\'t set value to email notification.');
```
Comment (akovalyov) [inline Service/NotificationManager.php:138]: should be in entity

File: Service/NotificationManager.php [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+      }
+
+      if (empty($value)) {
+        throw new InvalidArgumentException('Value can\'t be empty.');
```
Comment (akovalyov) [inline Service/NotificationManager.php:142]: you haven't specified a structure of value anywhere. Additionally, this should be in entity.

**Review by akovalyov — CHANGES_REQUESTED** [review]

File: Entity/NotificationMessageInterface.php (hunk 5 above)
Comment (groovili) [reply Entity/NotificationMessageInterface.php:7]: ok, agree

File: Entity/NotificationMessageInterface.php (hunk 6 above)
Comment (groovili) [reply Entity/NotificationMessageInterface.php:25]: Just title for the push notification.

File: Event/NotificationEventInterface.php (hunk 7 above)
Comment (groovili) [reply Event/NotificationEventInterface.php:10]: it's original naming from author, but I can fix that, no prob.

File: Service/NotificationManager.php (hunk 12 above)
Comment (groovili) [reply Service/NotificationManager.php:26]: ok

**Pull request closed without merging**